File: packages/adapters/android/ApplicationStateAdapter.kt
package os.awake.adapters.android

import java.security.MessageDigest

/**
 * STEP 1 SKELETON — Android foreground-application Event collector.
 *
 * Polls UsageStatsManager.queryEvents(...) (MOVE_TO_FOREGROUND / MOVE_TO_BACKGROUND)
 * and emits `ApplicationStateChanged` Events (schema 1.0.0).
 *
 * Invariants:
 *  - I-02: emits only the observed foreground/background transition.
 *  - I-09: raw package name is salted-hashed here and never leaves in the clear;
 *    buffer is drained locally by the bridge, no network.
 */
class ApplicationStateAdapter(
    /** Per-install salt from the Android keystore. Never synced. */
    private val installSalt: ByteArray,
    private val clock: () -> Long = System::currentTimeMillis,
    private val newUuid: () -> String = { java.util.UUID.randomUUID().toString() },
) {
    private val pending = ArrayDeque<NormalizedEvent>()

    /** @param rawEventType one of "MOVE_TO_FOREGROUND", "MOVE_TO_BACKGROUND", "TERMINATED". */
    fun onUsageEvent(packageName: String, rawEventType: String, timestampMs: Long) {
        val state = when (rawEventType) {
            "MOVE_TO_FOREGROUND" -> "Foreground"
            "MOVE_TO_BACKGROUND" -> "Background"
            "TERMINATED" -> "Terminated"
            else -> return
        }
        val hash = hashPackage(packageName)
        pending.addLast(
            NormalizedEvent(
                id = newUuid(),
                occurredAt = timestampMs.takeIf { it > 0 } ?: clock(),
                type = "ApplicationStateChanged",
                sourceType = "System",
                sourceId = "android.usage-events",
                subjectType = "Application",
                subjectId = hash,
                payload = mapOf("state" to state, "packageNameHash" to hash),
                schemaVersion = "1.0.0",
            ),
        )
    }

    fun drainPendingEvents(): List<NormalizedEvent> {
        val out = pending.toList()
        pending.clear()
        return out
    }

    private fun hashPackage(pkg: String): String {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(installSalt)
        val digest = md.digest(pkg.toByteArray(Charsets.UTF_8))
        return "sha256:" + android.util.Base64.encodeToString(
            digest,
            android.util.Base64.URL_SAFE or android.util.Base64.NO_PADDING or android.util.Base64.NO_WRAP,
        )
    }
}

File: packages/adapters/ios/ApplicationStateAdapter.swift
import CryptoKit
import Foundation

/// STEP 1 SKELETON — iOS foreground-application Event collector.
///
/// On iOS a general app cannot see *other* apps' foreground state without a
/// screen-time entitlement (DeviceActivity / FamilyControls). Where that is
/// granted, `DeviceActivityMonitor` callbacks feed this adapter; otherwise it
/// only reports transitions of the host app itself.
///
/// Invariants:
///  - I-02: emits only the observed foreground/background transition.
///  - I-09: bundle id is salted-hashed here; buffer drained locally, no network.
final class ApplicationStateAdapter {
    private var pending: [NormalizedEvent] = []
    private let installSalt: Data
    private let clockMs: () -> Int64
    private let newUuid: () -> String

    init(
        installSalt: Data,
        clockMs: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) },
        newUuid: @escaping () -> String = { UUID().uuidString }
    ) {
        self.installSalt = installSalt
        self.clockMs = clockMs
        self.newUuid = newUuid
    }

    /// `state` is one of: "Foreground", "Background", "Terminated".
    func onApplicationTransition(bundleId: String, state: String, timestampMs: Int64?) {
        guard ["Foreground", "Background", "Terminated"].contains(state) else { return }
        let hash = hashBundleId(bundleId)
        pending.append(
            NormalizedEvent(
                id: newUuid(),
                occurredAt: timestampMs ?? clockMs(),
                type: "ApplicationStateChanged",
                sourceType: "System",
                sourceId: "ios.app-lifecycle",
                subjectType: "Application",
                subjectId: hash,
                payload: ["state": state, "packageNameHash": hash],
                schemaVersion: "1.0.0"
            )
        )
    }

    func drainPendingEvents() -> [NormalizedEvent] {
        let out = pending
        pending.removeAll(keepingCapacity: true)
        return out
    }

    private func hashBundleId(_ bundleId: String) -> String {
        var hasher = SHA256()
        hasher.update(data: installSalt)
        hasher.update(data: Data(bundleId.utf8))
        let digest = hasher.finalize()
        let b64 = Data(digest).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
        return "sha256:\(b64)"
    }
}
